### DISCUSSION TITLE
Consider migrating to crossws

### DISCUSSION THREAD

**Discussion by tobiasdiez** [enisdenjo/graphql-ws#548] [General]
[crossws](https://github.com/unjs/crossws) is a new effort by the nuxt team to implement websocket servers across different platforms. Based on a quick look at the code, it should be relatively straightforward to migrate from the custom adapters in `src/use` to crossws. That would allow you to focus more on the graphql interface and mostly ignore the underlying websocket implementation.

(You probably want to wait a little bit until crossws' api gets more stable.)

**Comment by enisdenjo** [comment]
crossws could be another `use/` adapter, but I wont add it since there is no reason to add it. The implementations work. Also, the lib stays zero-dep.

**Comment by tobiasdiez** [comment]
> crossws could be another `use/` adapter

That is my strategy in https://github.com/apollo-server-integrations/apollo-server-integration-h3/pull/65.

> I wont add it since there is no reason to add it. The implementations work.

Right, the current implementation here seems stable. There are a few performance benefits of the design of crossws (eg avoiding heavy per-connection events, see "An API designed for speed" under https://bun.sh/docs/api/websockets#start-a-websocket-server). Moreover, in some of the implementations you have to keep track of all clients, which has additional memory costs and problems when scaling horizontally, https://github.com/enisdenjo/graphql-ws/blob/6013eb54829b27bd7c598f0985ec80a0e1acf09c/src/use/bun.ts#L77.

But I agree that currently there is not too much benefit. Perhaps the additional cloudflare adapter in crossws is the biggest plus (and I'm currently looking into adding azure web pubsub there). I just wanted to let you know about crossws and trust your judgment 😸

**Reply by enisdenjo** [reply]
Great work! You're very welcome to PR the adapter!
